core/common/src/main/java/alluxio/util/ShellUtils.java

```diff
@@ -193,6 +194,112 @@ private String run() throws ExitCodeException, IOException {
     }
+
+    /**
+     * Runs a command and returns its stdout, stderr and exit code.
+     * No matter it succeeds or not.
+     *
+     * @return {@link CommandReturn} object representation of stdout, stderr and exit code
+     */
+    protected CommandReturn runTolerateFailure() throws IOException {
+      Process process = new ProcessBuilder(mCommand).start();
+      CommandReturn cr = null;
+
+      try (BufferedReader inReader =
+                   new BufferedReader(new InputStreamReader(process.getInputStream()));
+           BufferedReader errReader =
+                   new BufferedReader(new InputStreamReader(process.getErrorStream()))) {
+        // read the output of the command
+        StringBuilder stdout = new StringBuilder();
+        String outLine = inReader.readLine();
+        while (outLine != null) {
+          stdout.append(outLine);
+          stdout.append("\n");
+          outLine = inReader.readLine();
+        }
+
+        // read the stderr of the command
+        StringBuilder stderr = new StringBuilder();
+        String errLine = errReader.readLine();
+        while (errLine != null) {
+          stderr.append(errLine);
+          stderr.append("\n");
+          errLine = inReader.readLine();
+        }
+
+        // wait for the process to finish and check the exit code
+        int exitCode = process.waitFor();
+        if (exitCode != 0) {
+          // log error instead of throwing exception
+          LOG.warn(String.format("Failed to run command %s%nExit Code: %s%nStderr: %s",
+                  Arrays.toString(mCommand), exitCode, stderr.toString()));
+        }
+
+        cr = new CommandReturn(exitCode, stdout.toString(), stderr.toString());
+
+        // destroy the process
+        if (process != null) {
+          process.destroy();
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException(e);
+      } finally {
+        if (process != null) {
+          process.destroy();
+        }
+      }
+      return cr;
+    }
+  }
+
+  @NotThreadSafe
+  private static class SshCommand extends Command {
+    protected String mHostName;
+
+    private SshCommand(String[] execString) {
+      this("localhost", execString);
+    }
+
+    private SshCommand(String hostname, String[] execString) {
+      super(execString);
+      mHostName = hostname;
+      mCommand = new String[]{"bash", "-c",
+              String.format("ssh %s %s %s", ShellUtils.COMMON_SSH_OPTS, hostname,
+                      String.join(" ", execString))};
+    }
+
+    protected String getHostName() {
+      return mHostName;
+    }
+  }
+
+  @NotThreadSafe
+  private static class ScpCommand extends Command {
+    protected String mHostName;
+
+    private ScpCommand(String remoteHost, String fromFile, String toFile) {
+      this(remoteHost, fromFile, toFile, false);
+    }
+
+    private ScpCommand(String remoteHost, String fromFile, String toFile, boolean isDir) {
```

Comment:
can you javadoc these parameters? (and the other constructor too)

Reply:
javadoc added after extracting these inner classes to separate class files and changed to `public`.